File: python/day_11/part_2.py
def solve(input: str) -> int:
    lines = [line.strip() for line in input.strip().split("\n") if line.strip()]

    graph = {}
    for line in lines:
        parts = line.split(":")
        device = parts[0].strip()
        outputs = parts[1].strip().split()
        graph[device] = outputs

    memo = {}

    def count_paths(node, has_dac, has_fft):
        if node == "out":
            return 1 if (has_dac and has_fft) else 0

        if (node, has_dac, has_fft) in memo:
            return memo[(node, has_dac, has_fft)]

        if node not in graph:
            return 0

        new_has_dac = has_dac or (node == "dac")
        new_has_fft = has_fft or (node == "fft")

        total = 0
        for neighbor in graph[node]:
            total += count_paths(neighbor, new_has_dac, new_has_fft)

        memo[(node, has_dac, has_fft)] = total
        return total

    return count_paths("svr", False, False)


if __name__ == "__main__":
    with open("inputs/day_11.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: go/day_11/part_2.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func solve(input string) int {
	lines := []string{}
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			lines = append(lines, line)
		}
	}

	graph := map[string][]string{}
	for _, line := range lines {
		parts := strings.Split(line, ":")
		device := strings.TrimSpace(parts[0])
		outputs := strings.Fields(strings.TrimSpace(parts[1]))
		graph[device] = outputs
	}

	memo := map[[3]interface{}]int{}

	var count_paths func(node string, has_dac bool, has_fft bool) int
	count_paths = func(node string, has_dac bool, has_fft bool) int {
		if node == "out" {
			if has_dac && has_fft {
				return 1
			}
			return 0
		}

		key := [3]interface{}{node, has_dac, has_fft}
		if val, ok := memo[key]; ok {
			return val
		}

		neighbors, ok := graph[node]
		if !ok {
			return 0
		}

		new_has_dac := has_dac || (node == "dac")
		new_has_fft := has_fft || (node == "fft")

		total := 0
		for _, neighbor := range neighbors {
			total += count_paths(neighbor, new_has_dac, new_has_fft)
		}

		memo[key] = total
		return total
	}

	return count_paths("svr", false, false)
}

func main() {
	input, err := os.ReadFile("../inputs/day_11.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
